### DISCUSSION TITLE
Eager load of "variants_including_master" doesn't affect "master" access

### DISCUSSION THREAD

**Discussion by crashtech** [solidusio/solidus#4561] [New Features or Ideas]
https://github.com/solidusio/solidus/blob/2b69316b670eb0913ff866eef1fb1f7fea3c8eb1/core/app/models/spree/product.rb#L34

Since the `master` method is direct connected to the association, there's no way take advantage of the eager loading. Which means:

```ruby
product = Spree::Product.includes(:variants_including_master).first
master = product.master # This will perform another query
```

The solution is pretty easy, I just don't have the time to read and follow the collaboration guidelines.

```ruby
# On core/app/models/spree/product.rb
def master
  return super unless variants_including_master.loaded?
  variants_including_master.find(&:is_master?)
end
```

**Comment by jhawthorn** [comment]
I don't want to do this because if will cause strange behaviour.

For example, the `master` association is `autosave: true`, but `variants_including_master` is not. So master is only autosaved if `variants_including_master` isn't loaded first.

Also if `master` is modified but not saved and then `variants_including_master loaded` loaded, those modifications will no longer be visible.
